httpx/errors/errors.go

```diff
@@ -199,3 +201,45 @@ func stacktrace(err error, skip int) errors.StackTrace {
 	}
 	return stack
 }
+
+// Will recover from a panic, and throw the error away.
+//
+// Useful for when you want to test that your panic handling is working
+// correctly. If your test actually throws a panic, it just crashes right
```

Comment:
The way this was worded, I thought it meant "if your test within its method body directly panics", but its more like "when your test triggers the code-under-test to panic" ... so, that makes sense, but then shouldn't the test be that one that has a `defer.IgnorePanic()` in its method body?

I saw the final version of the r101-experiments PR and just seemed odd to have `IgnorePanic` in the production code; I was originally thinking `PushPanicToChannel` should be a "terminal" handler and just suppress/not rethrow the panic, but AFAIU it could still rethrow the panic and "be fine", but then tests would have to have their own `defer IgnorePanic()`, but that seems fine/little more intuitive to me?

...or I suppose then the test might need to trigger the panic-causing before within an inner block/func so that it can recover that inner function, and then still have assertions run in its main block...

Reply:
The problem was that I didn't have access to the actual goroutine launching code.

It was like:

```
func SomeExportedFunction(...) {
  ...
  go privateFunction1(..., errChan);
  go privateFunction2(..., errChan);
  go privateFunction3(..., errChan);
  ...
}
```

And I wanted to test that SomeExportedFunction's goroutines handle errors correctly.

The problem with that is that go's testing libraries seem to count any unhandled panic in any goroutine as a test failure, so having `PushPanicToChannel` rethrow the error causes that kind of test to always fail. This is kind of annoying, since that's not what happens outside of a test, it would just kill the goroutine and whatever happens happens. It's frustrating we can't test that behaviour easily.

> but AFAIU it could still rethrow the panic and "be fine", but then tests would have to have their own `defer IgnorePanic()`

In my testing of this, that won't fix it. There's still a goroutine that hits the end with a panic (the `go someprivateFunction(...)` calls), and as far as I can tell go's testing suite makes that an automatic test failure. The panic doesn't make it's way back to the code starting the test.

We could have `PushPanicToChannel` be terminal, but it seemed like the intention with these is that you can chain them together, which seemed sensible / nice to me. And the only time that system breaks is when testing panic handling.

Maybe we need two versions like:
```
PushPanicToChannel
vs
PushPanicToChannelAndRecover
```

Still can feel the abstraction leaking, but I guess maybe that's clearer? I wish there was just a way to turn that 'any-panic-means-test-failure' feature off for a specific test.